### DISCUSSION TITLE
blake3 vs. fletcher4

### DISCUSSION THREAD

**Discussion by mabod** [openzfs/zfs#15421] [General]
Hello,

I am trying to understand the benefits of blake3 vs. fletcher4 checksums. But I dont get it. 
What is a good reason to move to blake3? Performance wise I believe fletcher4 is still faster. Is that right?

**Comment by IvanVolosyuk** [comment]
The fletcher4 checksum should be the fastest by far. I believe the only
reason to switch to a different checksum algorithm is if you use encryption
or dedup. I might be wrong though. ;)

**Comment by robn** [comment]
Blake3 is a cryptographic hash. Unless you need cryptographically-strong checksums (either you’re using dedup, or you have a particular application that requires it), then the default fletcher4 is still faster.

**Reply by rincebrain** [reply]
nopwrite is the reason I use edonr (or, eventually when I run 2.2, blake3, probably) on my things.

It's a very specialized use case, but when it's helpful, it's really helpful.

**Reply by deajan** [reply]
@rincebrain Why does only ednor provide nopwrite ? Is that still true with zfs 2.3.0 ? I'm searching for ways to optimize KVM Images, hence my question. Thanks.

**Reply by IvanVolosyuk** [reply]
Nopwrite is enabled for checksums which are sufficiently strong:
https://github.com/openzfs/zfs/blob/master/module/zfs/zio_checksum.c#L189
According to this sources: sha256, sha512, shein, edonr, blake3 have nopwrite.
I would imagine rincebrain used edonr as fastest of those with the constraints of the ZFS version preferred by him at that point. On ZFS 2.2 blake3 may be the fastest of strong checksums.
The builtin benchmark for the checksums is here on Linux systems:
```/proc/spl/kstat/zfs/chksum_bench```

**Reply by dodexahedron** [reply]
And, for what it's worth, on CPUs from the past decade, SHA512 tends to be quite a bit faster than SHA256 while not costing you anything except in the edge case that those extra 32 bytes would have made the difference between a record getting inlined in the dnode or not.

And blake3 is significantly faster than either of those (often multiple times faster), on x86 and 64-bit ARM variants, with the gains being more apparent with increasing size of input to it (so, bigger records benefit even more).

Regardless, hashing throughput is ***extremely*** unlikely to be a bottleneck or even more than a rounding error in the performance of ZFS on any system. Any amount of compression of any algorithm will cost a lot more than what the hash algorithm costs, in terms of CPU load, yet that'd also be an unwise thing to disable in most cases.

**Reply by IvanVolosyuk** [reply]
The last part is interesting. How can I actually measure the compression / hashing throughput in ZFS? I did a bit on benchmarking using command line tool for blake3, lz4 and fletcher4 (ripped AVX512f version from ZFS). I've got some numbers like
```
fletcher 30GB/s
lz4 5GB/s (compress)
blake3 3GB/s
```
Obviously, I didn't measured the same code we have in ZFS, but it gave me rough idea about the performance numbers.

There is also consideration of compressed ARC. Every time I want to access compressed data in ARC it will be decompressed, but hashing happens only once when data is read from disk.

**Reply by LittleNewton** [reply]
@IvanVolosyuk 
Could you please share your CPU information with me? I have recompiled OpenSSL with the flag `-march=native`, and the result showed that SHA-256 is much faster than SHA-512. My host is a Hyper-V Debian virtual machine based on an Intel Core 12700K CPU.

**Reply by IvanVolosyuk** [reply]
I have AMD 7950x3d, but how is related? I didn't check sha{256,512} peformance, just hashing and compression.

**Reply by LunarLambda** [reply]
Worth noting that the 12700K disables AVX512 depending on chip serial and microcode version, and AMD Zen 4 architecture only used a 256-bit datapath for AVX512. So benchmark results could vary strongly here.

In any case it would be good to take note of which extensions are being used for which hash in this discussion.

**Reply by dodexahedron** [reply]
Speaking of AVX512...

Also worth noting, for skylake chips, that there are two variants recognized by clang and gcc, with their own profiles - `skylake` and `skylake-x`. In my experience, detection of the correct profile for the actual hardware, when using -mtune=native, has been unreliable, with it often just picking skylake for machines that have CPUs that should be skylake-x. The biggest difference is the skylake-x chips have support for AVX512, whereas the non-x skylake chips only have AVX2 (that and -x are xeons and have a different cache layout).

Basically any Xeons in the Skylake family you should tell it skylake-x if it is detecting it as skylake only.

It could matter even more for several models of skylake chips in the Xeon-W line as well as all of the scalable gold and platinum lines, as those have 2 FMA units per core, while bronze and silver scalable skylake xeons only have 1 FMA unit per core.

Those are all older chips than those 12700K chips (same generation as the 6000 series core chips).

So yeah, what @LunarLambda said, but in a broader sense and going several generations back. Sometimes. On some chips. Because Intel.
